### PR TITLE
improved type-inference of union types

### PR DESCRIPTION
**Pull request by vtjnash**
This makes better use of Union types in call-sites by trying harder to form leaftype, and removing the hack from reflection. This should make it easier to add new inference heuristics. Needed to reduce `tupletype_len` back to a reasonable value so that many of the (sub)array-based tests don't require hours to finish inferring, now that inference can see into some of their calls better and it starts trying to infer a couple of complex recursive functions (used for indexing?).

### CODE REVIEW

**Comment by TotalVerb** [comment]
Might be wise to alert the StaticArray maintainers of this change:

@andyferris @c42f

File: base/inference.jl [hunk 1]
```diff
@@ -2497,12 +2545,14 @@ function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller
...
     if frame === nothing
         code.inInference = true
-        frame = InferenceState(code, true, true, caller.params) # always optimize and cache edge targets
+        frame = InferenceState(code, #=optimize=#true, #=cached=#true, caller.params) # always optimize and cache edge targets
         if frame === nothing
             code.inInference = false
             return Any, nothing
         end
-        frame.parent = caller
+        if caller.cached # don't involved uncached functions in cycle resolution
```
Comment (tkelman) [inline base/inference.jl]: don't involve ?

**Comment by andyferris** [comment]
I appreciate the heads-up. :)

I'm not going to pretend to understand the details (yet), but I will flag the concern that a tuple length of 8 would be smaller than a 3x3 matrix necessary to do fast geometric stuff with StaticArrays, so I'd better check the benchmarks/codegen. I'm not sure if the copious amounts of generated functions more-or-less avoid this heuristic. Could someone remind me what the exact meaning of `tupletype_len` is (i.e. at precisely what points widening might occur)?

I'm assuming that the change of `tupletype_len` *won't* be backported to v0.6? (Also, could it be argued that the improvements to inference and the change in heuristics are in-fact separate changes?).

Finally, I'm also a little concerned with things like `TypedTables` where having only a small number of types is a bit prohibitive already with a value of 16.  I've been working on packages (some internal) that use only lispy-recursion and *no* generated functions to great effect to deal with hetrogenously-typed collections (codegen is often awesome - thanks guys!) so I'd be sad to see this work become less useful for v0.6-final (it would also be bad for my JuliaCon presentation...). 

I still remember the time when @timholy wanted it changed to 42. :)

**Comment by c42f** [comment]
Thanks for the heads up!  Would we be insulated from some of these changes if `NTuple` was dealt with as a special case? We only need homogenously typed tuples, after all.

**Comment by andyferris** [comment]
Also, how does `MAX_TUPLE_SPLAT` relate to `MAX_TUPLETYPE_LEN`?

File: base/inference.jl [hunk 2]
```diff
@@ -1493,34 +1533,37 @@ function abstract_apply(af::ANY, fargs::Vector{Any}, aargtypes::Vector{Any}, vty
...
     ctypes = Any[Any[]]
     for i = 1:nargs
         if aargtypes[i] === Any
             # bail out completely and infer as f(::Any...)
-            # instead could keep what we got so far and just append a Vararg{Any} (by just
-            # using the normal logic from below), but that makes the time of the subarray
-            # test explode
-            ctypes = Any[Any[Vararg{Any}]]
+            # instead could infer the precise types for the types up to this point and just append a Vararg{Any}
+            # (by just using the normal logic from below), but that makes the time of the subarray test explode
```
Comment (martinholters) [inline base/inference.jl]: Have you checked whether that is still an issue?

**Comment by martinholters** [comment]
Could you give an example of an improvement obtained by this? And could that then be made a test?

**Comment by KristofferC** [comment]
Benchmark from https://github.com/KristofferC/Tensors.jl at https://gist.github.com/KristofferC/cf1da0501c8b554cc64869e81cf877cf. Only the AD part was affected which does end up with quite large nested tuples.

**Comment by mbauman** [comment]
The (sub)array infrastructure was developed specifically against the constraints of inference at the time.  If those constraints are changing, maybe that's where changes should occur.  Happy to discuss how/why things there work the way they do.

The other place tuple lengths have been important in the past is in number of arguments for indexing into multidimensional arrays.  Is that still the case?  8 is likely too small — it means that getindex would only be efficient for 6-dimensional arrays and setindex! for 5.

**Comment by vtjnash** [comment]
s/don't require hours to finish inferring/would never have finished inference/

Reducing the value caused enough of an impact on the inference quality to make it through most of the tests, but the unsoundness in the termination / recursion detection is a severe blocker for any future improvements to inference. New PR coming.

**Pull request closed without merging**